host/sgx/linux/vdso.c

```diff
@@ -0,0 +1,259 @@
+OE_NEVER_INLINE
+oe_result_t oe_vdso_enter(
+    void* tcs,
+    uint64_t arg1,
+    uint64_t arg2,
+    uint64_t* arg3,
+    uint64_t* arg4,
+    oe_enclave_t* enclave)
+{
+    oe_ecall_context_t ecall_context = {0};
+    oe_result_t result = OE_UNEXPECTED;
+    struct sgx_enclave_run run = {0};
+    vdso_args_t return_args = {0};
+    int return_value = 0;
+    uint64_t rflags = 0;
+    uint32_t mxcsr = 0;
+    uint16_t fcw = 0;
+    uint64_t rbx = 0;
+    uint64_t r12 = 0;
+    uint64_t r13 = 0;
+    uint64_t r14 = 0;
+    uint64_t r15 = 0;
+    uint64_t rsp = 0;
+
+    oe_setup_ecall_context(&ecall_context);
+
+    run.tcs = (uint64_t)tcs;
+    run.user_handler = (uint64_t)oe_vdso_user_handler;
+    run.user_data = (uint64_t)&return_args;
+
+    /* Record the current RSP */
+    asm volatile("mov %%rsp, %0" : "=r"(rsp));
+
+    while (1)
+    {
+        /* Compiler will usually handle this on exiting a function that uses
+         * AVX, but we need to avoid the AVX-SSE transition penalty here
+         * manually as part of the transition to enclave. See
+         * https://software.intel.com/content/www/us/en/develop/articles
+         * /avoiding-avx-sse-transition-penalties.html */
+        if (oe_is_avx_enabled)
+            OE_VZEROUPPER;
+
+        /* Save and restore MXCSR, x87 control word, RFLAGS, and non-volatile
+         * registers (excpet for RSP and RBP that are expected to be preserved)
+         * before and after the vDSO call conform the Linux x64 ABI */
+        asm volatile("stmxcsr %[mxcsr] \n\t"  // Save MXCSR
+                     "fstcw %[fcw] \n\t"      // Save x87 control word
+                     "mov %%rbx, %[rbx] \n\t" // Save RBX
+                     "mov %%r12, %[r12] \n\t" // Save R12
+                     "mov %%r13, %[r13] \n\t" // Save R13
+                     "mov %%r14, %[r14] \n\t" // Save R14
+                     "mov %%r15, %[r15] \n\t" // Save R15
+                     // Temporarily set the rsp to the end of the &rflags
+                     // so that we can use pushfq to save RFLAGS into rflags
+                     "mov %[rflags_addr_end], %%rsp \n\t"
+                     "pushfq \n\t"
+                     "mov %[rsp], %%rsp \n\t" // Restore the RSP
+                     : [rbx] "=m"(rbx),
+                       [r12] "=m"(r12),
+                       [r13] "=m"(r13),
+                       [r14] "=m"(r14),
+                       [r15] "=m"(r15)
+                     : [fcw] "m"(fcw),
+                       [mxcsr] "m"(mxcsr),
+                       [rflags_addr_end] "r"((uint64_t)&rflags + 8),
+                       [rsp] "m"(rsp)
+                     :);
+
+        return_value = (_vdso_sgx_enter_enclave)(
+            arg1,
+            arg2,
+            (uint64_t)&ecall_context,
+            ENCLU_EENTER,
+            0 /* r8 */,
+            0 /* r9 */,
+            &run);
+
+        asm volatile("mov %[rflags_addr_start], %%rsp \n\t" // Temproily set RSP
```

Comment:
This may work, but is not reliable. Between the call in line 203 and this block in 212, the compiler could emit code that uses any register.

Reply:
updated